flink-core/src/main/java/org/apache/flink/core/state/StateFutureImpl.java

```diff
@@ -61,23 +66,34 @@ public <U> StateFuture<U> thenApply(Function<? super T, ? extends U> fn) {
                 return StateFutureUtils.completedFuture(r);
             } else {
                 StateFutureImpl<U> ret = makeNewStateFuture();
-                completableFuture.thenAccept(
-                        (t) -> {
-                            callbackRunner.submit(
-                                    () -> {
-                                        ret.complete(fn.apply(t));
-                                        callbackFinished();
-                                    });
-                        });
+                completableFuture
+                        .thenAccept(
+                                (t) -> {
+                                    callbackRunner.submit(
+                                            () -> {
+                                                ret.complete(fn.apply(t));
+                                                callbackFinished();
+                                            });
+                                })
+                        .exceptionally(
+                                (e) -> {
+                                    exceptionHandler.handleException(
```

Comment:
shall we allow the handler return something? if the execption is recoveryable, we can continue the operation.

Reply:
Please note that the `exceptionHandler` here is for non-recoverable errors by framework. As for the recoverable ones (like state access timeout or something), there will be automatically retry in StateExecutor. In this layer, we only focus on the future-related stuff.